src/Nancy.Hosting.Self/NancyHost.cs

```diff
@@ -124,20 +125,27 @@ public void Start()
         {
             this.StartListener();
 
-            Task.Factory.StartNew(async () =>
+            Task.Run(() =>
             {
-                try
+                var semaphore = new Semaphore(this.configuration.MaximumConnectionCount, this.configuration.MaximumConnectionCount);
+                while (!this.stop)
                 {
-                    while(!this.stop)
+                    semaphore.WaitOne();
+
+                    this.listener.GetContextAsync().ContinueWith(async (contextTask) =>
                     {
-                        HttpListenerContext context = await this.listener.GetContextAsync().ConfigureAwait(false);
-                        await this.Process(context).ConfigureAwait(false);
-                    }
-                }
-                catch(Exception ex)
-                {
-                    this.configuration.UnhandledExceptionCallback.Invoke(ex);
-                    throw;
+                        try
+                        {
+                            semaphore.Release();

```

Comment:
I think this is a valid concern, however if we were to move the Release to a finally clause, we won't be able to accept new requests until the full request is handled, rather than being able to continue accepting new requests after the context is retrieved. The options for finally clauses would in my mind be: 1. Put a finally clause on the try block inside of the continuation, which wouldn't help if GetContextAsync() were to throw, or put a new try finally around this.listener.GetContextAsync(), which wouldn't really work as we want that to execute as a hot task so we wouldn't be blocking the semaphore. 

Another point I would like to make is that an exception here is likely due to a configuration issue, and is going to bubble out as an unhandled exception, so most likely the application is going to terminate unless someone writes explicit handling around the host start and tries to recover. Here's the exceptions that potentially could come out of the BeginGetContext that GetContextAsync invokes: https://docs.microsoft.com/en-us/dotnet/api/system.net.httplistener.begingetcontext?view=netframework-4.7 . I guess the big question to answer here is: Should an exception thrown from GetContextAsync() be recoverable? I think maybe if the answer to that is yes, the application's scope may be better suited utilizing a different host.

I do think the point is fair though and worth discussion!